### PR TITLE
Custom button data

### PR DESCRIPTION
**Pull request by Mayo19**

### CODE REVIEW

**Comment by chelliah** [comment]
screen shot 2018-05-08 at 6 10 01 pm

**Comment by chelliah** [comment]
screen shot 2018-05-09 at 4 55 56 pm
@trisweb great call on the alignment

File: src/editors/button/ButtonEditor.js [hunk 1]
```diff
@@ -164,6 +183,18 @@ export default class ButtonEditor extends React.Component {
     };
     buttonAttrs.style = getButtonStyleString(borderRadius, padding, fontSize, width);
 
+    if (trackEvent && eventName) {
+      buttonAttrs['data-attrs-event'] = JSON.stringify({event: eventName}).replace(/\"/g, "&quot;");
+    }
+
+    if (trackEvent && eventName) {
+      buttonAttrs['data-attrs-event'] = JSON.stringify({event: eventName}).replace(/\"/g, "&quot;");
+    }
```
Comment (rayd) [inline src/editors/button/ButtonEditor.js]: oops, looks like a duplicate line

File: src/editors/button/ButtonEditor.js [hunk 2]
```diff
@@ -164,6 +183,18 @@ export default class ButtonEditor extends React.Component {
...
+
+    if (updateUserProperties && userPropertiesToUpdate) {
```
Comment (rayd) [inline src/editors/button/ButtonEditor.js]: May also want to check that `userPropertiesToUpdate` has at least one key

File: src/editor-actions/ButtonAction.js [hunk 3]
```diff
@@ -47,87 +84,195 @@ export default class ButtonAction extends React.Component {
...
+
     const hasMoreThanOneStep = numPages > 1;
+    const remainingUserProperties = userPropertiesDropdown.filter((property) => userPropertiesToUpdate.get(property.value) == undefined);
+    const exampleInput = keyToAdd && userPropertiesDropdown.find((property) => property.value == keyToAdd).options[0].name;
+    const inputFieldType = this.getUserPropertyType(keyToAdd, userPropertiesDropdown);
 
     const dropdownNodes = isActive ? (
-      <Menu style={dropdownStyles}>
-
+      <Menu style={{...dropdownStyles, ...flexColumn, maxHeight: 300, overflow: 'scroll' }}>
```
[the remark this answers is not in the document]
Comment (chelliah) [reply src/editor-actions/ButtonAction.js]: yeahhhh this was the hardest part of the pr haha, this menu has to be capped at a max height, and setting any overflow rules will cause the dropdown to be clipped. Escaping overflows is impossible of any parent is positioned (in this case, the button)

File: src/components/DropDownMenu.js [hunk 4]
```diff
@@ -211,8 +222,14 @@ export default class DropDownMenu extends React.Component {
...
                 <span>{ selectedOption ? selectedOption.label : defaultValue }</span>
-                <span style={{display: 'inline-flex', position: 'absolute', right: '0px', height: '20px'}}>
-                  <DownButton iconStyle={{color: 'grey'}}/>
+                <span style={{
+                  display: 'inline-flex',
+                  right: '0px',
+                  height: '20px',
+                  position: 'absolute',
+                  cursor: 'pointer'
+                }}>
```
[the remark this answers is not in the document]
Comment (chelliah) [reply src/components/DropDownMenu.js]: good call

File: src/editor-actions/ButtonAction.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rayd) [reply src/editor-actions/ButtonAction.js]: Hmm, so what if we make the dropdowns have `display: none` when hidden and `display: flex` when showing? and then we remove the `max-height` on that panel? I guess the panel would grow extra long if they added a lot of properties, but they'd have to add a lot to get themselves in trouble...
deepinscreenshot_select-area_20180511111126

[the remark this answers is not in the document]
Comment (chelliah) [reply src/editor-actions/ButtonAction.js]: yeah I was afraid of screen overflow, not sure which is worse.

[the remark this answers is not in the document]
Comment (rayd) [reply src/editor-actions/ButtonAction.js]: I'd argue that having to scroll down inside the panel to access all of the options of the dropdown is probably worse...

File: src/editors/button/ButtonEditor.js [hunk 5]
```diff
@@ -187,11 +187,7 @@ export default class ButtonEditor extends React.Component {
       buttonAttrs['data-attrs-event'] = JSON.stringify({event: eventName}).replace(/\"/g, "&quot;");
     }
 
-    if (trackEvent && eventName) {
-      buttonAttrs['data-attrs-event'] = JSON.stringify({event: eventName}).replace(/\"/g, "&quot;");
-    }
-
-    if (updateUserProperties && userPropertiesToUpdate) {
+    if (updateUserProperties && Object.keys(userPropertiesToUpdate) > 0) {
```
Comment (rayd) [inline src/editors/button/ButtonEditor.js]: Whoops, looks like we're missing a ```.length``` here

**Pull request merged**